src/controllers/midi/midimessage.h

```diff
@@ -204,13 +212,32 @@ struct MidiInputMapping {
     }
 
     bool operator==(const MidiInputMapping& other) const {
-        return key == other.key && options == other.options &&
-                control == other.control && description == other.description;
+        return std::visit([&](auto first, auto second) {
+            using T = std::decay_t<decltype(first)>;
+            using U = std::decay_t<decltype(second)>;
+
+            if constexpr (!std::is_same_v<T, U>) {
+                return false;
+            } else if constexpr (std::is_same_v<T, ConfigKey>) {
+                return key == other.key &&
+                        options == other.options &&
+                        std::get<ConfigKey>(control) == std::get<ConfigKey>(other.control) &&
+                        description == other.description;
+            } else if constexpr (std::is_same_v<T, QJSValue>) {
+                return key == other.key && options == other.options &&
+                        std::get<QJSValue>(control).strictlyEquals(
+                                std::get<QJSValue>(other.control)) &&
+                        description == other.description;
+            } else
+                static_assert(always_false_v<T>, "non-exhaustive visitor");
```

Comment:
shouldn't this also suffice? I don't want to introduce unnecessary boilerplate.
```suggestion
                static_assert(false, "non-exhaustive visitor");
```

Reply:
This line comes from [`std::visit`](https://en.cppreference.com/w/cpp/utility/variant/visit#Example) documentation. Coming from Python and Kotlin, I admit I'm having troubles to understand C++'s type system.

Reply:
yeah I saw that it comes from there, I'm just confused whether its really required. 

> I admit I'm having troubles to understand C++'s type system.

Yeah... its complicated. I can recommend you the cppcon "back to basics" talks. Probably "Templates - Bob Steagall", maybe "Function and Class Templates - Dan Saks" as primer. Also look into overload resolution since that becomes important for generics (I think). I'm don't understand it good enough either, just barely good enough for basic everyday use.